File: sdk-py/examples/basic.py
"""Two-agent demo for the AgentMailbox Python SDK.

Mirrors examples/basic.ts. Run after starting the AgentMailbox server:

    cd ~/agentmailbox && npm run start
    python sdk-py/examples/basic.py
"""

from __future__ import annotations

import asyncio
import os

from agentmailbox import AgentMailbox


SERVER = os.environ.get("AGENTSMCP_SERVER", "http://localhost:3000")


async def main() -> None:
    # ResearchAgent
    async with AgentMailbox("researcher@demo", server=SERVER) as researcher:
        await researcher.connect()
        result = await researcher.send(
            "writer@demo",
            {"task": "summarize diffusion models", "papers": ["paper1", "paper2"]},
            context_snapshot={"step": "research_complete", "paper_count": 2},
        )
        thread_id = result.thread_id
        print(f"[researcher] sent on thread: {thread_id}")

    # WriterAgent — cold start, picks up full context
    async with AgentMailbox("writer@demo", server=SERVER) as writer:
        await writer.connect()
        inbound = await writer.receive()
        print(f"[writer] unread messages: {len(inbound.messages)}")
        print(f"[writer] context.snapshot: {inbound.context.snapshot}")
        assert inbound.context.snapshot == {
            "step": "research_complete",
            "paper_count": 2,
        }, "writer should pick up researcher's snapshot cold"

        reply = await writer.send(
            "researcher@demo",
            {"draft": "Diffusion models work by..."},
            thread_id=thread_id,
            context_snapshot={"step": "draft_complete", "word_count": 500},
        )
        print(f"[writer] reply sent: {reply.message_id}")
        await writer.mark_read(thread_id)

    # Researcher syncs updated context
    async with AgentMailbox("researcher@demo", server=SERVER) as researcher:
        await researcher.connect()
        context = await researcher.sync(thread_id)
        print(f"[researcher] synced snapshot: {context.snapshot}")
        assert context.snapshot == {"step": "draft_complete", "word_count": 500}
        print(
            "[researcher] recent payloads:",
            [m.payload for m in context.recent_messages],
        )


if __name__ == "__main__":
    asyncio.run(main())

File: examples/basic.ts
import { AgentMailbox } from "../src/agentmailbox";

async function main() {
  const server = process.env.AGENTMAILBOX_SERVER ?? "http://localhost:3000";

  // ResearchAgent
  const researcher = new AgentMailbox({
    agentId: "researcher@demo",
    server,
  });
  await researcher.connect();

  const { threadId } = await researcher.send(
    "writer@demo",
    { task: "summarize diffusion models", papers: ["paper1", "paper2"] },
    { contextSnapshot: { step: "research_complete", paperCount: 2 } }
  );
  console.log("[researcher] sent on thread:", threadId);

  // WriterAgent — picks up full context even after restart
  const writer = new AgentMailbox({
    agentId: "writer@demo",
    server,
  });
  await writer.connect();

  const { messages, context } = await writer.receive();
  console.log("[writer] unread messages:", messages.length);
  console.log("[writer] context.snapshot:", context.snapshot);
  // → { step: "research_complete", paperCount: 2 }
  // Writer is fully caught up. Never started cold.

  const reply = await writer.send(
    "researcher@demo",
    { draft: "Diffusion models work by..." },
    {
      threadId,
      contextSnapshot: { step: "draft_complete", wordCount: 500 },
    }
  );
  console.log("[writer] reply sent:", reply.messageId);

  await writer.markRead(threadId);

  // Researcher syncs full thread
  const synced = await researcher.sync(threadId);
  console.log("[researcher] synced snapshot:", synced.context.snapshot);
  console.log(
    "[researcher] recent messages:",
    synced.context.recentMessages.map((m) => ({ from: m.from, payload: m.payload }))
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
